File: Coin Change[11-August-2023]/Memoization/Solution.java
// >> JAVA CODE <<


class Solution {
    public long solve(int i,int curSum,int[] coins,int reqsum,long[][] dp)
    {
        if(curSum==reqsum) return 1;
        if(i==coins.length) return 0;

        if(dp[i][curSum]!=-1) return dp[i][curSum];

        if(curSum+coins[i]<=reqsum)
            return dp[i][curSum]=solve(i,curSum+coins[i],coins,reqsum,dp) + solve(i+1,curSum,coins,reqsum,dp);
        
        return dp[i][curSum]=solve(i+1,curSum,coins,reqsum,dp);
    }
    public long count(int coins[], int N, int sum) {
        long[][] dp=new long[N][sum];
        
        for(int i=0;i<N;i++) Arrays.fill(dp[i],-1);
        
        return solve(0,0,coins,sum,dp);
    }
}

File: Coin Change[11-August-2023]/Memoization/Solution.cpp
//  >> C++ CODE <<


class Solution {
  public:
    long long int solve(int i,int curSum,int coins[],int N,int sum,vector<vector<long>>& dp)
    {
        if(curSum==sum) return 1L;
        if(i==N) return 0L;
        
        if(dp[i][curSum]!=-1) return dp[i][curSum];
        
        if(curSum+coins[i]<=sum)
            return dp[i][curSum]=solve(i+1,curSum,coins,N,sum,dp) + solve(i,curSum+coins[i],coins,N,sum,dp);
            
        return dp[i][curSum]=solve(i+1,curSum,coins,N,sum,dp);
    }
    long long int count(int coins[], int N, int sum) {

        vector<vector<long>> dp(N,vector<long>(sum,-1));

        return solve(0,0,coins,N,sum,dp);
    }
};
